lib/std/heap/general_purpose_allocator.zig

```diff
@@ -227,6 +228,7 @@ pub fn GeneralPurposeAllocator(comptime config: Config) type {
             }
         };
         const LargeAllocTable = std.AutoHashMapUnmanaged(usize, LargeAlloc);
+        const AllocSizeTable = std.AutoHashMapUnmanaged(usize, usize);
```

Comment:
If we're trying to catch invalid frees, wouldn't it also be pertinent to check for correct alignment as well? Passing allocating `[]T`, and then trying to free as `[]u8` instead of `[]align(@alignOf(T)) u8` is a semi-common bug out in the while, and a lot harder to spot than invalid length.

Using a simple type like `struct { size: usize, align_log2: u8 }` would suffice.

Reply:
Makes sense. To be clear, `[]u8` ptr is still pointing to the correct memory block in this case, right (otherwise the current patch would catch this). Is it sufficient to catch just the wrong alignment or should we also ensure the freed type matches the allocated type?

Reply:
There wouldn't be a way to check what the allocated type is, because the implementation only ever allocations, resizes, and frees, of bytes. Alignment is a property of memory, not of the type.

Reply:
Why would the alignment need to be stored in addition to the address?

Reply:
@Vexu the allocator interface contract (the comments in `std.mem.Allocator.VTable`) require that an allocation pointer that is resized or freed (as in, is passed to the `resize` or `free` methods of the `VTable`) must be passed with the same length and alignment value as was initially requested. This PR makes GPA store a hash map that maps an allocation address to its requested size and alignment values, in order to check that when it is resized/freed, that it is with the correct values.